### ISSUE TITLE
 User data was not in the MIME multipart format with module >20.2.1

### ISSUE THREAD

**Issue by maruina** [terraform-aws-modules/terraform-aws-eks#2955]
## Description

Creating a new EKS cluster with a Managed NodeGroup fails with the following error
```
│ Error: waiting for EKS Node Group (green:kube_system-20240305170030320200000001) create: unexpected state 'CREATE_FAILED', wanted target 'ACTIVE'. last error: : Ec2LaunchTemplateInvalidConfiguration: User data was not in the MIME multipart format.
```

- [ ] ✋ I have searched the open/closed issues and my issue is not listed.

## Versions

- Module version [Required]: `>20.2.1`

- Terraform version:
```
➜ terraform -version
Terraform v1.6.1
on darwin_arm64
+ provider registry.terraform.io/hashicorp/aws v5.39.1
+ provider registry.terraform.io/hashicorp/cloudinit v2.3.3
+ provider registry.terraform.io/hashicorp/helm v2.12.1
+ provider registry.terraform.io/hashicorp/kubernetes v2.26.0
+ provider registry.terraform.io/hashicorp/time v0.10.0
+ provider registry.terraform.io/hashicorp/tls v4.0.5
```

## Reproduction Code [Required]
```
module "eks" {
  source  = "terraform-aws-modules/eks/aws"
  version = "~> 20.0"

  cluster_name    = "green"
  cluster_version = "1.29"
  tags = {
    "karpenter.sh/discovery" : "green",
  }

  # Access config
  authentication_mode = "API_AND_CONFIG_MAP"
  # Allow account-admin to be cluster admin
  access_entries = {
    account_admin = {
      principal_arn = "arn:aws:iam::111111111111:role/aws-reserved/sso.amazonaws.com/foo"
      policy_associations = {
        compute_admin = {
          policy_arn = "arn:aws:eks::aws:cluster-access-policy/AmazonEKSClusterAdminPolicy"
          access_scope = {
            type = "cluster"
          }
        }
      }
    }
    compute_admin = {
      principal_arn = "arn:aws:iam::111111111:role/aws-reserved/sso.amazonaws.com/bar"
      policy_associations = {
        cluster_admin = {
          policy_arn = "arn:aws:eks::aws:cluster-access-policy/AmazonEKSClusterAdminPolicy"
          access_scope = {
            type = "cluster"
          }
        }
      }
    }
  }

  enable_irsa = true

  vpc_id     = module.vpc.vpc_id
  subnet_ids = module.vpc.private_subnets

  cluster_endpoint_public_access       = false
  cluster_endpoint_private_access      = true

  cluster_service_ipv4_cidr = "172.17.0.0/16"

  cluster_addons = {
    coredns = {
      most_recent = true
    }
    kube-proxy = {
      most_recent = true
    }
    vpc-cni = {
      most_recent    = true
      before_compute = true
      configuration_values = jsonencode({
        env = {
          # Reference docs https://docs.aws.amazon.com/eks/latest/userguide/cni-increase-ip-addresses.html
          ENABLE_PREFIX_DELEGATION = "true"
          WARM_PREFIX_TARGET       = "1"
        }
      })
    }
  }

  eks_managed_node_groups = {
    kube_system = {
      min_size     = 1
      max_size     = 5
      desired_size = 2

      instance_types = ["m5.large"]
      taints = [
        {
          key    = "CriticalAddonsOnly"
          value  = "true"
          effect = "NO_SCHEDULE"
        }
      ]
      labels = {
        "role" : "worker"
      }
    }
  }
}
```

`terraform apply` for the example above fails, while it is successful when the module is pinned to 
```
version = "20.2.1"
```

If I look at the AWS console I can see `Health Issue 1` for the MNG, and the error is the same from Terraform
Not using workspace, I'm creating a new cluster from scratch with a new project.

## Expected behavior

The MNG should report 0 Health Issue and nodes are created

## Actual behavior

The MNG reports an health issue

### Terminal Output Screenshot(s)

<img width="1493" alt="image" src="https://github.com/terraform-aws-modules/terraform-aws-eks/assets/4821896/f9e3af91-4fab-44cd-ad4e-92ab7ff862c6">

**Comment by bryantbiggs** [comment]
would you mind copy+pasting the rendered user data from the instance (please be sure to remove any sensitive/identifiable details)?

**Comment by maruina** [comment]
> would you mind copy+pasting the rendered user data from the instance (please be sure to remove any sensitive/identifiable details)?

I'm not 100% sure how to do this :)
I have 0 instances coming up, and I couldn't find the user-data in the LaunchTemplate.

**Comment by bryantbiggs** [comment]
if you go into the AWS console -> EC2 - select one of the instances that was created by the managed nodegroup (it will create instances, they just are failing to join the cluster/control plane) -> Select the `Actions` dropdown menu -> Select `Instance settings` -> Select `Edit user data` 

That will bring up a window that shows the rendered user data in a greyed out box. Just copy+remove any sensitive deets+paste that 
<img width="474" alt="image" src="https://github.com/terraform-aws-modules/terraform-aws-eks/assets/10913471/86442424-9dbc-4c03-80e6-3b35d36d370e">

**Comment by maruina** [comment]
A working node has this userdata
```
MIME-Version: 1.0
Content-Type: multipart/mixed; boundary="//"

--//
Content-Type: text/x-shellscript; charset="us-ascii"
#!/bin/bash
set -ex
B64_CLUSTER_CA=XXXXXXX
API_SERVER_URL=https://YYYYYYY.gr7.xx-yyyy-1.eks.amazonaws.com
K8S_CLUSTER_DNS_IP=172.17.0.10
/etc/eks/bootstrap.sh green --kubelet-extra-args '--node-labels=eks.amazonaws.com/nodegroup-image=ami-0134cdd1bd3d81c83,eks.amazonaws.com/capacityType=ON_DEMAND,eks.amazonaws.com/sourceLaunchTemplateVersion=2,eks.amazonaws.com/nodegroup=kube_system-XXXX,role=worker,eks.amazonaws.com/sourceLaunchTemplateId=lt-0f82261ef5efXXXX --register-with-taints=CriticalAddonsOnly=true:NoSchedule --max-pods=110' --b64-cluster-ca $B64_CLUSTER_CA --apiserver-endpoint $API_SERVER_URL --dns-cluster-ip $K8S_CLUSTER_DNS_IP --use-max-pods false

--//--
```
and the `userdata` in its Launch Template is the same.

When using the latest module, I can't get an EC2 instance up and running. I have this error
<img width="1150" alt="Screenshot 2024-03-06 at 10 49 18" src="https://github.com/terraform-aws-modules/terraform-aws-eks/assets/4821896/a5e34fb9-b8c3-4252-899d-81754af47f3b">
and when looking at the LT I see the following userdata
<img width="894" alt="Screenshot 2024-03-06 at 10 51 04" src="https://github.com/terraform-aws-modules/terraform-aws-eks/assets/4821896/29096bb5-2074-4ef1-b96c-680d02140408">

Maybe it's missing some parameters?

**Comment by anatolijd** [comment]
i'm on version 20.5.2, and can confirm this.

When variable `cluster_service_ipv4_cidr`   is explicitly set (even to default value "10.100.0.0.0/16") , then LaunchTemplate modification fails:

apply log:
```
  # module.eks.module.eks_managed_node_group["one"].aws_launch_template.this[0] will be updated in-place
  ~ resource "aws_launch_template" "this" {
      ~ default_version         = 1 -> (known after apply)
        id                      = "lt-004855e1f"
      ~ latest_version          = 1 -> (known after apply)
        name                    = "one-20240308111207051100000001"
        tags                    = {
            "Jira" = "RFPD-41956"
        }
      + user_data               = "ZXhwb3J0IFNFUlZJQ0VfSVBWNF9DSURSPTEwLjEwMC4wLjAvMTYK"
        # (9 unchanged attributes hidden)

        # (5 unchanged blocks hidden)
    }

│ Error: waiting for EKS Node Group (rancher-upgrade:node-group-1-20240308111208240100000003) version update (50abed5f-793b-3eea-b9ca-c6113bbeda51): unexpected state 'Failed', wanted target 'Successful'. last error: : Ec2LaunchTemplateInvalidConfiguration: User data was not in the MIME multipart format.
│ 
│   with module.eks.module.eks_managed_node_group["one"].aws_eks_node_group.this[0],
│   on .terraform/modules/eks/modules/eks-managed-node-group/main.tf line 341, in resource "aws_eks_node_group" "this":
│  341: resource "aws_eks_node_group" "this" {
```  

It seems user_data value is wrong:
```
$ echo ZXhwb3J0IFNFUlZJQ0VfSVBWNF9DSURSPTEwLjEwMS4wLjAvMTYK | base64 -d
export SERVICE_IPV4_CIDR=10.101.0.0/16
```

It is neither MIME-enveloped  nor having  #cloud-config or #!/bin/sh on first line.

**Comment by antonbabenko** [comment]
This issue has been resolved in version 20.8.3 :tada:
